File: design_patterns/C++/Structural/facade.cxx
#include <iostream>

class Computer {
  public:
	void getElectricShock() {
		std::cout << "Ouch!" << std::endl;
	}

	void makeSound() {
		std::cout << "Beep beep!" << std::endl;
	}

	void showLoadingScreen() {
		std::cout << "Loading..." << std::endl;
	}

	void bam() {
		std::cout << "Ready to be used!" << std::endl;
	}

	void closeEverything() {
		std::cout << "Bup bup bup buzzzz!" << std::endl;
	}

	void sooth() {
		std::cout << "Zzzzz" << std::endl;
	}

	void pullCurrent() {
		std::cout << "Haaah!" << std::endl;
	}
};

class ComputerFacade {
  protected:
	Computer* computer = nullptr;

  public:
	ComputerFacade(Computer* computer) {
		this->computer = computer;
	}

	void turnOn() {
		this->computer->getElectricShock();
		this->computer->makeSound();
		this->computer->showLoadingScreen();
		this->computer->bam();
	}

	void turnOff() {
		this->computer->closeEverything();
		this->computer->pullCurrent();
		this->computer->sooth();
	}
};

int main() {
	auto computer = new ComputerFacade(new Computer());
	computer->turnOn();
	// =================
	// Ouch!
	// Beep beep!
	// Loading...
	// Ready to be used!
	// =================
	computer->turnOff();
	// ===================
	// Bup bup bup buzzzz!
	// Haaah!
	// Zzzzz
	// ===================
	return 0;
}

File: design_patterns/Python/Structural/facade.py
class Computer:

    def get_electric_shock(self):
        print('Ouch!')

    def make_sound(self):
        print('Beep beep!')

    def show_loading_screen(self):
        print('Loading...')

    def bam(self):
        print('Ready to be used!')

    def close_everything(self):
        print('Bup bup bup buzzzz!')

    def sooth(self):
        print('Zzz...')

    def pull_current(self):
        print('Haaaah!')


class ComputerFacade:

    def __init__(self, computer):
        self.computer = computer

    def turn_on(self):
        self.computer.get_electric_shock()
        self.computer.make_sound()
        self.computer.show_loading_screen()
        self.computer.bam()

    def turn_off(self):
        self.computer.close_everything()
        self.computer.pull_current()
        self.computer.sooth()


if __name__ == '__main__':

    computer = ComputerFacade(Computer())

    computer.turn_on()
    # Ouch!
    # Beep beep!
    # Loading...
    # Ready to be used!

    computer.turn_off()
    # Bup bup bup buzzzz!
    # Haaaah!
    # Zzz...
